packages/simcore-sdk/src/simcore_sdk/node_ports_common/aws_s3_cli.py

```diff
@@ -0,0 +1,307 @@
+        for filename in filenames:
+            file_path = Path(dirpath) / filename
+            total_size += _get_size_and_manage_symlink(Path(file_path))
+    return ByteSize(total_size)
+
+
+async def _sync_sources(
+    aws_s3_cli_settings: AwsS3CliSettings,
+    progress_bar: ProgressBarData,
+    *,
+    source: str,
+    destination: str,
+    local_dir: Path,
+    exclude_patterns: set[str] | None,
+    debug_logs: bool = True,
+) -> None:
+
+    if source.startswith("s3://"):
+        folder_size: ByteSize = await _get_s3_folder_size(
+            aws_s3_cli_settings, s3_path=shlex.quote(source)
+        )
+    else:
+        folder_size = await _get_local_folder_size(Path(source))
+
+    cli_command = [
+        "aws",
+        "s3",
+        "sync",
+        "--delete",
+        shlex.quote(source),  # <-- source
+        shlex.quote(destination),  # <-- destination
+        # filter options
+        *_get_exclude_filters(exclude_patterns),
+        "--no-follow-symlinks",
+    ]
+
+    if not aws_s3_cli_settings.AWS_S3_CLI_PROVIDER.value == S3Provider.AWS:
+        cli_command.insert(
+            1, f"--endpoint-url {aws_s3_cli_settings.AWS_S3_CLI_S3.S3_ENDPOINT}"
+        )
+
+    async with progress_bar.sub_progress(
+        steps=folder_size,
+        progress_unit="Byte",
+        description=f"transferring {local_dir.name}",
+    ) as sub_progress:
+        aws_s3_cli_log_parsers: list[BaseLogParser] = (
+            [DebugLogParser()] if debug_logs else []
+        )
+        aws_s3_cli_log_parsers.append(SyncAwsCliS3ProgressLogParser(sub_progress))
+
+    await _async_aws_cli_command(
+        *cli_command,
+        aws_s3_cli_settings=aws_s3_cli_settings,
+        aws_cli_s3_log_parsers=aws_s3_cli_log_parsers,
+    )
+
+
+def _raise_if_directory_is_file(local_directory_path: Path) -> None:
+    if local_directory_path.exists() and local_directory_path.is_file():
+        raise AwsS3CliDirectoryNotFoundError(local_directory_path=local_directory_path)
+
+
+async def sync_local_to_s3(
+    aws_s3_cli_settings: AwsS3CliSettings,
+    progress_bar: ProgressBarData,
+    *,
+    local_directory_path: Path,
+    upload_s3_link: AnyUrl,
+    exclude_patterns: set[str] | None = None,
+    debug_logs: bool = False,
+) -> None:
+    """transfer the contents of a local directory to an s3 path
+
+    :raises e: AwsS3CliFailedError
+    """
+    _raise_if_directory_is_file(local_directory_path)
+
+    upload_s3_path = upload_s3_link
+    _logger.debug(" %s; %s", f"{upload_s3_link=}", f"{upload_s3_path=}")
+
+    await _sync_sources(
+        aws_s3_cli_settings,
+        progress_bar,
+        source=f"{local_directory_path}",
+        destination=f"{upload_s3_path}",
+        local_dir=local_directory_path,
+        exclude_patterns=exclude_patterns,
+        debug_logs=debug_logs,
+    )
+
+    # Remove the temporary created .osparclink files after they were uploaded to S3
+    for textfile_path in local_directory_path.rglob(
```

Comment:
It might be a good idea to use a context manager for these special links. this way you are sure this never get skipped.

Reply:
I would also expect a context manager that wraps the call to `_sync_sources`:
- `convert_symlinks` context manger: takes care of creating symlink files and removing them when the upload to s3 is over
- `restore_symlinks` context manger that recreates the symlinks after the download and removes the files
As mentioned by Sylvain that's a good way to avoid these always run regardless of the errors that the sync may raise (because it will raise errors in the future)